### PR TITLE
[RDY] Fixes for the new TUI

### LINKED ISSUE #2000: New terminal UI broke bracketed paste mode
With https://github.com/neovim/neovim/pull/1820, [bracketed paste](http://cirw.in/blog/bracketed-paste) no longer works, [ConradIrwin/vim-bracketed-paste](https://github.com/ConradIrwin/vim-bracketed-paste) has stopped working.

Neovim (`66a03a7`)

screen shot 2015-02-17 at 9 10 56

Vim `7.4.488`

screen shot 2015-02-17 at 9 12 02

### PR DESCRIPTION
**Pull request by tarruda**
Fixes #2000, #2003, #2005 and #2013

### CODE REVIEW

**Comment by justinmk** [comment]
See https://github.com/neovim/neovim/pull/1053 for an existing PR that may be useful to implement bracketed paste mode.

It would also be nice to publish `PastePre` and `PastePost` event as discussed there.

**Comment by justinmk** [comment]
FYI, clang caught some things: http://neovim.org/doc/reports/clang/

**Comment by tarruda** [comment]
> See #1053 for an existing PR that may be useful to implement bracketed paste mode.

Thats helpful, thanks

> It would also be nice to publish PastePre and PastePost event as discussed there. GUIs can publish these events to get the same behavior.

It seems to me that the `Paste{Pre,Post}` autocommands only have the purpose of setting/unsetting the `paste` option. If so it might be simpler to hardcode that action directly in C. The [doc](http://neovim.org/doc/user/options.html#'paste') even says that this is handled automatically in the GUI version.

**Comment by justinmk** [comment]
Using the Paste events for bracketed mode was mostly secondary. I think in general having ClipboardPaste/Yank events will be useful. But it's not urgent right now.

**Comment by tarruda** [comment]
The last commit tries to implement support bracketed paste. After some basic tests it seems to be working. @renstrom I assume you use this feature a lot, can you give this PR a try?

File: src/nvim/tui/term_input.inl [hunk 1]
```diff
@@ -176,6 +177,29 @@ static void read_cb(RStream *rstream, void *rstream_data, bool eof)
...
+          || !strncmp(ptr, "\x1b[201~", 6))) {
+      bool enable = ptr[4] == '0';
+      // Advance past the sequence
+      rbuffer_consumed(input->read_buffer, 6);
+      if (enable) {
+        // Get the current mode
+        int state = get_real_state();
+        if (state & NORMAL) {
+          // Enter insert mode
+          input_enqueue(cstr_as_string("i"));
```
Comment (justinmk) [inline src/nvim/tui/term_input.inl]: `input_enqueue()` ignores mappings, right?

**Comment by renstrom** [comment]
@tarruda just tried this branch and it seems like it's still not working, tried with and without [ConradIrwin/vim-bracketed-paste](https://github.com/ConradIrwin/vim-bracketed-paste) enabled, though I assume the preferred way is without it.

This is the `.vimrc` I used:

```vim
all plug#begin()
Plug 'tpope/vim-sensible'
Plug 'sheerun/vim-polyglot'
call plug#end()
```

Also, don't know if it matters but it was over an SSH connection. Is there any more detailed info I can provide?

**Comment by elmart** [comment]
@tarruda Please add fixes here for generated clang-analysis warnings (there are corresponding new coverity issues too, that should be gone with them).

**Comment by tarruda** [comment]
> @tarruda Please add fixes here for generated clang-analysis warnings (there are corresponding new coverity issues too, that should be gone with them).

:+1:

**Comment by tarruda** [comment]
> Also, don't know if it matters but it was over an SSH connection. Is there any more detailed info I can provide?

It shouldn't matter if its over SSH. Are you using it under tmux? I've been unable to make it work for tmux despite wrapping the bracketed paste enable sequence in `\ePtmux;`.

In any case, could you try it again? I've fixed it to use the `pastetoggle` option

**Comment by renstrom** [comment]
>Are you using it under tmux?

Nope. I wanted to make sure it worked normally first, but my intent was to try tmux after.

>In any case, could you try it again? I've fixed it to use the `pastetoggle` option

Indeed it seems to be fixed now! :) However as you said, I couldn't get it working in tmux either.

**Comment by tarruda** [comment]
@renstrom It seems tmux doesn't require any wrapper sequence for enabling bracketed paste like the vim-bracketed-paste plugin sends. Should be working now.

**Comment by renstrom** [comment]
@tarruda Awesome! Just verified it myself as well.

**Comment by tarruda** [comment]
Since #1997 will remove all termcap options, the t_SI/tEI hack for changing cursor shape wont work. To fill the gap, I've added support for iterm/xterm(v 282+) sequences that can be disabled by setting the `NVIM_TUI_DISABLE_CURSOR_SHAPE` environment variable.

@Tranquility can you test if this branch works as expected for you?

**Comment by tarruda** [comment]
> In the commit message, it's "use after free" not the other way round.

:+1:

**Comment by Tranquility** [comment]
It looks like everything is working for me, but shouldn't the default be that the cursor shape changes?

**Comment by tarruda** [comment]
> It looks like everything is working for me, but shouldn't the default be that the cursor shape changes?

It has been brought to my attention that some terminals may display unrecognized codes, and since not every terminal support this feature I though it was better to leave it as opt-in.

File: src/nvim/tui/tui.c [hunk 2]
```diff
@@ -724,23 +764,6 @@ static void flush_buf(UI *ui)
   data->bufpos = 0;
 }
 
-static char *get_term_option(UI *ui, char *option)
-{
-  TUIData *data = ui->data;
-
-  char *rv = pmap_get(cstr_t)(data->option_cache, option);
```
Comment (oni-link) [inline src/nvim/tui/tui.c]: `option_cache` was not removed. Will it be used in the future?

Comment (tarruda) [inline src/nvim/tui/tui.c]: I don't think so, will remove it on #2025. thanks

**Comment by PixelPirate** [comment]
Can someone elaborate how the `iterm/xterm(v 282+) sequences` workaround for `t_SI/tEI` is supposed to work?

I used to use:
```
if &term =~ '^xterm'
  let &t_SI .= "\<Esc>[6 q"  " solid vertical line in insert mode
  let &t_EI .= "\<Esc>[2 q"  " solid block in normal mode
endif
```

What should I use now? (Sorry if this is an inappropriate question in this discussion)

**Pull request merged**